### DISCUSSION TITLE
Cannot clear storage after updating to 3.7.2

### DISCUSSION THREAD

**Discussion by sandergo90** [pmndrs/zustand#1775] [Q&A]
Hi,

After updating to version 3.7.2, the clearStorage function doesn't seem to work correctly anymore. It looks like PR #894 broke it for us.  When resetting the first change in https://github.com/pmndrs/zustand/pull/894/files, it works again. We have the following code:

```
export function StoreProvider(props: { children: ReactNode }) {
  const { children } = props;

  const hasHydrated = useStore(state => state._hasHydrated);

  // Clear persisted storage when new app install
  const initializeApp = async () => {
    const appInstalled = await AsyncStorageService.getItem(APP_INSTALLED_KEY);

    if (!appInstalled) {
      useStore.persist.clearStorage();
      useStore.persist.rehydrate();
    }
  };

  useEffect(() => {
    initializeApp();
  }, []);

  return hasHydrated ? <Provider createStore={() => useStore}>{children}</Provider> : null;
}
```

It's a provider that wil check if the app was installed. If opening the app, and the appInstalled state Is not true, it would clear the storage because we don't want the persisted storage to still be here when you deleted the application.

Sander

**Comment by dai-shi** [comment]
I'm not sure if I follow precisely, but can you work around with `options.merge`? Do you have any other suggestions?

**Comment by byteab** [comment]
so you mean that this line
```useStore.persist.clearStorage();```
doesn't clear the storage ? right ?

**Comment by Mugilan-Codes** [comment]
clearStorage() is not working as expected

**Comment by victorkvarghese** [comment]
## Summary

App is not resetting the store after calling clearStorage. If i provide timeout , things are working. I need to know if this is the right way to clear storage and reset the store. 

On log out i'm clearing the store and reloading the app. At first i thought store was not cleared because app didn't restart. So added RNRestart. but still things are not working. With timeout it works. 

Do i need to manually reset each slice , even after clearing the storage?


## Link to reproduction

```
  const onLogOut = async () => {
    RNQueryClient.clear();
    RNQueryClient.invalidateQueries();
    useCustomStore.persist.clearStorage();
    useCustomStore.persist.rehydrate();
    RNRestart.restart();
  };
```

If i add timeout then it works,

```
  const onLogOut = async () => {
    RNQueryClient.clear();
    RNQueryClient.invalidateQueries();
    useCustomStore.persist.clearStorage();
    setTimeout(() => {
      useCustomStore.persist.rehydrate();
      RNRestart.restart();
    }, 400);
  };
```

## Version
React Native : 
```
 "zustand": "^4.4.4"
 "react-native": "0.72.6",
```

**Reply by victorkvarghese** [reply]
@dai-shi can someone help? It used work with older versions though (without timeout)

**Reply by dai-shi** [reply]
Can you identify which version exactly changed the behavior?

**Reply by victorkvarghese** [reply]
@dai-shi I need to check again, but i think it was working on v4.4.2 or v4.4.3 

Do i need to do something after clearing storage ? Like rehydrate or reload the app ? I tried reload using RNRestart , but it didnt work. So had to use timeout

**Reply by dai-shi** [reply]
Knowing which version works and which version doesn't (more precisely which PR it is) might reveal the behavior.

**Reply by MaxRobertsDear** [reply]
@victorkvarghese I think that your mental model needs a little tweaking. 

Understandable to assume that `useStore.persist.clearStorage();` would also clear the cache, but it's just syntactic sugar for removing said item from your persisted store. 

For example, if you were using the [SecureStore](https://docs.expo.dev/versions/latest/sdk/securestore/) `useStore.persist.clearStorage();` would simply invoke `SecureStore.deleteItemAsync(“nameOfKey”)`.

When does the persisted store come into play? on initialisation (well, _creation_) of the cached store: store is initiated with default initial state and then is hydrated from persisted store (assuming an async persisted store - for sync persisted store, cached store is created with persisted data if present). 

Anytime the cached store is updated, so is the persisted store. 

The consuming component is reading from the cache, not the persisted store. 

If you clear the persisted storage, your cache is unaffected. In the example of the timeout working, the whole app is being restated RNRestart.restart() - thereby clearing the cache and re-initialising. The catch here is that doing this would also add the initialised cached store to the persisted store. 

### Overview for useStore.persis.clearStorage()
#### clearStorage 
- calls storage?.removeItem("nameOfStore")
- does not interact with hasHydrated flag 
- does not change cached state of store
#### hasHydrated
- boolean that shows whether the store has been hydrated from storage
- set to true on successful hydration 


### Resetting the (cached) store
Unless you actually care about deleting the persisted store, it sounds like you want to reset the cache instead. [These docs](https://docs.pmnd.rs/zustand/guides/how-to-reset-state) provide a couple of approaches. Note that changes to the cached store will be reflected in the persisted store.

I hope that the above is helpful! Happy to articulate differently if it's not clear 😆
